### ISSUE TITLE
Unicode supplementary character in CSS content value causes parse error

### ISSUE THREAD

**Issue by skyfaller** [validator/validator#1344]
URL being validated or code to reproduce error:
```css
/* decorative text in hr at top & bottom of page */
.menu hr::after,
footer hr::after {
	background: var(--background);
	content: '🚧';
	padding: 0 .2em;
	position: relative;
	top: -.8em;
}
```
You can see this in context at: https://www.maximumethics.dev/css/style.css

```
Error: content: Parse Error.
```
If it matters, [this CSS validates on the old school validator](https://jigsaw.w3.org/css-validator/validator?uri=https%3A%2F%2Fwww.maximumethics.dev%2Fcss%2Fstyle.css&profile=css3svg&usermedium=all&warning=1&vextwarning=&lang=en), but [not on Nu](https://validator.w3.org/nu/?showsource=yes&doc=https%3A%2F%2Fwww.maximumethics.dev%2Fcss%2Fstyle.css).

Is it really not sufficient to [serve the CSS file as UTF-8](https://stackoverflow.com/questions/10393462/placing-unicode-character-in-css-content-value), as I'm doing? If it really [must be escaped Unicode](https://stackoverflow.com/questions/190396/adding-html-entities-using-css-content), then I apologize and you can close this bug.

**Comment by sideshowbarker** [comment]
> If it matters, [this CSS validates on the old school validator](https://jigsaw.w3.org/css-validator/validator?uri=https%3A%2F%2Fwww.maximumethics.dev%2Fcss%2Fstyle.css&profile=css3svg&usermedium=all&warning=1&vextwarning=&lang=en), but [not on Nu](https://validator.w3.org/nu/?showsource=yes&doc=https%3A%2F%2Fwww.maximumethics.dev%2Fcss%2Fstyle.css).

Yeah, this is a bug in the Nu checker, then. I’ll make time to investigate

> Is it really not sufficient to [serve the CSS file as UTF-8](https://stackoverflow.com/questions/10393462/placing-unicode-character-in-css-content-value), as I'm doing?

As far as I understand, yeah, it should be

cc @ylafon

**Comment by sideshowbarker** [comment]
OK after doing some troubleshooting on this, I see that it occurs only for code points above U+FFFF (where most emoji are) but not for code points below U+FFFF.

So, for example, this will be cause no parse error:

```
content: '⭐'
```
…because the code point of that `⭐` character is U+2B50.

But this will cause a parse error:

```
 content: '𐀀'
```
…because the code point of that `𐀀` character is U+10000.

And the thing that’s special about all code points above U+FFFF is that in Java they’re represented as a [surrogate pair](https://stackoverflow.com/questions/5903008/what-is-a-surrogate-pair-in-java) of two [char](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html) values, while everything below U+FFFF is represented by a single char value.

And so the bug here seems to be that somewhere in the HTML checker code where we construct the stylesheet contents to pass to the CSS checker backend, we’re somehow causing those surrogate pairs to not get passed on correctly.

I’ll need to do more investigation to figure out where that might be happening.

**Comment by sideshowbarker** [comment]
@skyfaller Can you please also try by going to https://jigsaw.w3.org/css-validator/#validate_by_input and pasting in the snippet from the issue description? (I mean rather than giving the https://www.maximumethics.dev/css/style.css URL to it)

I think in that case you should also get a parse error (I do). And if so, that would indicate there’s in fact a problem in the CSS-validator backend that we need to figure out and fix (and not actually in the HTML checker code).

**Comment by sideshowbarker** [comment]
Nevermind about my previous comment — direct input works as expected there as long as it includes an `@charset`.

In the mean time, at https://github.com/w3c/css-validator/pull/382 I’ve opened a PR which, if it gets approved and merged, will fix this.

**Comment by sideshowbarker** [comment]
@skyfaller Thanks much for catching this and reporting it — I’ve pushed the fix to https://validator.w3.org/nu/
